programs/cli_wallet/main.cpp

```diff
@@ -279,6 +279,19 @@ int main( int argc, char** argv )
       {
          wallet_cli->register_api( wapi );
          wallet_cli->start();
+
+         fc::set_signal_handler([&wallet_cli, &wapi, &wallet_file](int signal) {
+            ilog( "Captured SIGINT not in daemon mode" );
+            wapi->save_wallet_file(wallet_file.generic_string());
+            wallet_cli->quit_blocked_thread();
+         }, SIGINT);
+
+         fc::set_signal_handler([&wallet_cli, &wapi, &wallet_file](int signal) {
+            ilog( "Captured SIGTERM not in daemon mode" );
+            wapi->save_wallet_file(wallet_file.generic_string());
+            wallet_cli->quit_blocked_thread();
```

Comment:
I found a simpler solution: just call ``fclose(stdin);``. Please test on Windows.

Reply:
Thanks ! Will check it ...

Reply:
@pmconrad I've checked your method on Ubuntu 18.04 and your solution is simpler than mine I absolutely agree with you but your solution has disadvantage:

after unlocking wallet if you press CTRL+C then you cannot see what you type in command line, you need to restart it, in my approach we don't have this issue.

Don't you like my approach with quitting `blocked thread` ?

Thanks !

Reply:
No, I don't like that. :-)
* The thread may not be actually blocked at the time the signal is received. In that case your code will segfault.
* When compiled with editline, there is a separate background thread active that will be left dangling with your method.

I could not reproduce the behaviour you described, but it might be a good idea to restore tty settings when cli shuts down.

Reply:
`cli_wallet` uses `future` and `promise` which has `blocked_thread` member, when `SIGINT` or `SIGTERM` is caught I use method `quit` of the `blocked_thread` member, thus we will stop `cli_wallet`'s thread.

@pmconrad 
1. What do you mean by "The thread may not be actually blocked at the time the signal is received" ?
2. My solution just `quit` only `cli_wallet`'s thread, do you mean my solution may `quit` another thread instead of required ?
3. To reproduce the behaviour I've described you should evaluate next 3 steps (tested on Ubuntu 18.04):
    3.1. unlock `cli_wallet`
    3.2  press `CTRL+C`
    3.3 try to write anything in command line tool and you won't see what are you typing there

Thanks !

Reply:
1. If the signal is received before you call wallet_cli->wait(), blocked_thread will be a nullptr. Admittedly this is unlikely in this specific case, but there may be other uses of fc::rpc::cli where it is less unlikely.
2. No, there *is* another thread related to cli that is not stopped by your solution.
3. I tried that (but not on Ubuntu). But never mind, if you see the behaviour then tty settings must be cleaned up. I suppose you do use editline? Can you reproduce this reliably?

Reply:
I've used `editline` through VS code and after new updates seems all works fine enough now with your solution @pmconrad (I mean `fclose(stdin)` ! Very simple and efficient ! Thank you very much ! Will do new commit now and will close next PR in `fc` library https://github.com/bitshares/bitshares-fc/pull/77 cause we don't need anymore